### PR TITLE
[test][sdpa] Cover explicit-stream and device-context boundaries

### PR DESCRIPTION
**Pull request by 0z5a**
**Incremental review:** [Files changed](https://github.com/NVIDIA/cudnn-frontend/pull/1149/files) against `develop`.

Refs #510

## Current-tree audit

- **Originally reported sites already protected: all of them.** `engines.py::_execute` and `_execute_by_tensor` wrap `_execute_resolved` (the only caller of `api.execute`) in `_torch_stream_context(stream, api.q_desc.device)`, so every adapter copy/zero/scale the issue lists sits inside that outer context even where the method has no local one. The SM100 FP8/MXFP8 arms and `_execute_dense_prepared` additionally carry local contexts, the latter with `verify_current=True` for an explicit stream.
- **Remaining gap: coverage, not a bug.** The two architecture-independent cases lived in `sdpa/frost/test_sdpa_stream_ordering.py`, whose module-level `requires_pre_rubin_blackwell` gate skips the whole file on every other part (5/5 skipped on an L20), and the helper boundaries (`verify_current`, the default-stream sentinels, exception/nesting restoration, device isolation) were not covered anywhere.
- **Why the existing explicit-stream and prepared tests do not cover it:** `test_fp8_dense_postprocessing_follows_explicit_stream` needs SM100 + the DSL, so it cannot execute here, and it exercises the kernel path rather than the helper contract.
- **Sentinel hypothesis:** the `0/1/2` early-return in `_torch_stream_context` is a suspicion, not a confirmed race. It is now probed behaviourally (no stream object constructed, ambient stream left in place) and **not changed**.

Per-site status table (`ALREADY_COVERED` / `COVERAGE_GAP` / `ARCH_UNVALIDATED`) and the caller chains are in the evidence bundle.

## Change

- **Regression coverage only.** New `test/python/sdpa/test_stream_context_contract.py` (13 cases, no Blackwell gate): enter/restore, exception restore, nesting/repetition, `verify_current` vs the raw-handle fast path, sentinels, an ordering probe on a real ambient-vs-launch mismatch, `_get_default_stream(None)`, two-device isolation, a fresh stream/device/storage set on every round, CUDA-graph capture with the inputs changed before the replay, the native dense f16 lane with the handle bound to a non-default stream (captured, replayed and compared against an independent float64 reference), and a structural pin of the engine lane's outer guard (removing it fails the test rather than only an architecture CI run).
- The one architecture-independent case moved out of the gated file; no duplicate node id, and the kernel-level cases keep their gate and their decorators.
- No FP8 epilogue work, no public API change, no added synchronization, no product code change.

## Evidence

- Base / head: `667fe4ce8ce437866217066f075fd4dcecad6eac` / `3b6b9da7`. `cudnn.__file__` verified to be the `wt-510` checkout (extension built from that tree's C++).
- Host: L20 (SM89), driver 595.91.07, cuDNN 9.26 runtime, torch 2.11.0+cu128.
- Executed ids and results: `sdpa/test_stream_context_contract.py` **10 passed** with two visible devices (the two-device case skips, and only that case, when one device is visible); the gated engine file collects 5 and skips 5 by its architecture gate; `test_import_boundaries.py` **15 passed**.
- Stream topology and sensitivity: the ordering probe keeps ambient ≠ launch, and a test-local mutation (`context -> nullcontext`) reads the poison from 8e8 spin cycles up while the real context is correct at 1e9 (per-spin results in the audit file).
- Architectures not tested: SM100/SM120 kernel paths — they need the Blackwell CI; this PR claims helper-layer evidence only.

## Follow-ups

- FP8 glue removal stays separate.
- Sentinel semantics are not changed without dedicated evidence.

## Summary by CodeRabbit

* **Tests**
  * Added CUDA-gated SDPA coverage for stream switching and restoration, current-stream resolution, device isolation, ordered operations, and execution safeguards.
  * Added validation for CUDA graph capture and replay, native execution results, and fresh stream bindings.
  * Added graceful handling for non-CUDA environments and configurations with fewer than two devices.
  * Removed a duplicate architecture-independent stream-context test from the Blackwell-specific test suite.

## Follow-up at `ebd29bbe`

Added the three L20-executable cases that were still missing, and corrected the spin comment to the measured behaviour. Two-device runs: **13 passed**; one card: **11 passed / 2 skipped** (the per-round and isolation cases skip alone, the rest still run).

| Layer | Evidence on L20 |
|---|---|
| Helper / binding semantics | 13 cases; mutation with the context replaced by a no-op fails the deterministic enter/restore assertion |
| Real kernel (native dense f16) | handle bound to a non-default stream while torch sits on another: max abs error **1.7e-4** against an independent float64 reference, **1.8e-4** after capture + replay with changed inputs |
| Value-race trial | with the context disabled the trial stays **masked** on this host (ambient work is scheduled after the launch spin); its docstring says so and the deterministic cases carry the detector role |
| FROST/DSL kernel rows | still skip on sm_89 (`test_sdpa_stream_ordering.py` 5/5, `test_sdpa_stream_respect.py` 3/3, frontend-integration kernel cases): architecture CI evidence required, not claimed here |

Re-verified at `3483f155`: both initialization-order findings from the next review are fixed -- the per-round case now records its producer fill and waits on it from the launch stream with a barrier before the trial, and the native case resets its output on the bound stream while `g.execute` runs with torch's ambient stream current; the capture/replay path orders its input update and output reset on the capture stream too. 28 passed (13 contract + 15 import-boundary).

### CODE REVIEW

**Review by YangXu1990uiuc — APPROVED** [review]
**Codex bot review · model: `gpt-6-astra`**

Reviewed `3b6b9da7938f` against base `667fe4ce8ce4`.

Thank you for your contribution! Moving the helper checks out of the Blackwell-only suite and covering restoration/device boundaries makes sense. I found one nonblocking test-reliability issue, independently confirming the existing poison-write ordering comment. With a controlled delay before the ambient-stream poison fill, the unmodified test fails despite the correct helper; explicitly completing initialization fixes that failure and still detects a deliberately disabled context. Please address the initialization ordering before relying on this as a CI regression detector.

- **[P2] Still open:** [Order input initialization before the launch-stream trial](https://github.com/NVIDIA/cudnn-frontend/pull/1149#discussion_r4052764560).

Validation: Exact-head normal pytest execution: all 10 new stream/device cases and 15 existing import-boundary cases pass (25 passed, no skips). L40S is the primary GPU and A100 the second visible GPU, so the device-isolation case executes. PyTorch 2.13.0+cu130, cuDNN 9.25.1; cudnn import path verified against this checkout and the native extension built from its sources. Four independent controlled ordering outcomes match expectations: ordinary test passes; delayed poison with the correct helper fails; ordered initialization with the correct helper passes; ordered initialization with a deliberately disabled context fails. This confirms the linked test defect and the proposed fix's sensitivity. The diff changes only tests. The engine adapter guard and production helper implementation are unchanged; kernel performance and CPU hot-path behavior are therefore outside this delta.

Limitations: Helper-layer validation does not establish the unchanged SM100/SM120 kernel paths. No new broad GPU CI or kernel benchmark was run. Current-head Style workflow 35434483248 is awaiting fork authorization.

Approved: no P0 or high-risk P1 found. Remaining findings stay with the owner; merge timing stays with the owner.

**Comment by YangXu1990uiuc** [comment]
The [Style run for `3b6b9da7938f`](https://github.com/NVIDIA/cudnn-frontend/actions/runs/35434483248) passed after maintainer authorization. The remaining initialization-ordering comment is covered in my review; I have not requested another broad GPU pipeline while that test is being adjusted.

Codex bot review — model `gpt-6-astra`.

**Review by YangXu1990uiuc — COMMENTED** [review]
**Codex bot review · model: `gpt-6-astra`**

Reviewed `d31cfdf1d12a` against base `667fe4ce8ce4`.

Thank you for your contribution! The new event fixes the ambient-stream poison race, and narrowing the claim about sleep-based detection is appropriate. One nonblocking part of the previous initialization comment remains: the initial arange/full writes are still on the caller/default stream, outside the new event dependency. Please order that stream before the ambient poison write, e.g. ambient.wait_stream(torch.cuda.current_stream(device)) immediately before entering the ambient context, or create both tensors inside that context before recording poison_ready. My existing approval stands.

- **[P2] Still open:** [Include default-stream tensor initialization in the event dependency](https://github.com/NVIDIA/cudnn-frontend/pull/1149#discussion_r4052764560).

Validation: Exact-head normal run: 25 passed (10 stream/device tests plus 15 import-boundary tests), with L40S and A100 visible so device isolation executes. cuDNN 9.25.1, PyTorch 2.13.0+cu130; source import and matching native-source identity verified. Independent controls on the unmodified test: ordinary execution passes; delayed ambient poison passes with the new event; delayed default-stream initialization fails despite the correct helper (64/64 mismatches, max absolute difference 133); ordering that initialization before the trial passes. The fix therefore makes progress but does not yet cover every initialization stream. Reviewed the full one-file incremental diff. No production helper, kernel, API or launch behavior changed.

Limitations: These results concern the helper test and its initialization reliability, not the unchanged Blackwell kernel paths. No broad GPU CI or performance benchmark was requested for this test-only follow-up. The exact-head Style check is awaiting maintainer authorization.

Comment only: this PR already has an active approval; preserving existing approvals.

File: test/python/sdpa/test_stream_context_contract.py [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+    from cudnn.sdpa.fwd.api_dsl import _torch_stream_context
+
+    device = torch.device("cuda")
+    ambient = torch.cuda.Stream(device=device)
+    launch = torch.cuda.Stream(device=device)
+    real = torch.arange(64, dtype=torch.float32, device=device)
+    x = torch.full((64,), -7.0, dtype=torch.float32, device=device)
+
+    with torch.cuda.stream(ambient):
+        x.fill_(-7.0)
```
[the remark this answers is not in the document]
Comment (0z5a) [reply test/python/sdpa/test_stream_context_contract.py:261]: Fixed in d31cfdf1: the poison fill records `poison_ready`, the launch stream waits on it before it spins and restores, and no device-wide synchronize is added.

Sensitivity was checked by replacing `_torch_stream_context` with a no-op in place: the deterministic enter/restore case fails as it should, while the value-race trial stays masked on this L20 because the ambient work is scheduled behind the launch spin here. The trial's docstring now says exactly that and leaves the kernel-level claim in the gated Blackwell file. 10 passed on two L20s at `d31cfdf1`.

**Comment by YangXu1990uiuc** [comment]
The [Style check for `d31cfdf1`](https://github.com/NVIDIA/cudnn-frontend/actions/runs/35485064200) passed after maintainer authorization. The remaining default-stream initialization dependency is described in the follow-up review; no additional broad GPU pipeline was requested for this test-only revision.

Codex bot review — model `gpt-6-astra`.

**Comment by 0z5a** [comment]
Extended at `ebd29bbe` while closing the remaining A2 rows: a fresh stream/device/storage set every round, CUDA-graph capture with the inputs changed before the replay, and the one dense lane this part can execute -- the native backend with the handle bound to a non-default stream, captured and replayed against an independent float64 reference (max abs error 1.7e-4 on both streams, 1.8e-4 after replay). 13 passed on two L20s; 11 passed / 2 skipped with one card. The FROST/DSL kernel rows still skip on sm_89 and stay with the architecture CI.

**Review by YangXu1990uiuc — COMMENTED** [review]
**Codex bot review · model: `gpt-6-astra`**

Reviewed `ebd29bbef4a5` against base `667fe4ce8ce4`.

Thank you for your contribution! The original ordering test now correctly completes initialization before its trial; my delayed-initialization reproduction passes. The expanded suite passes 28 cases locally. Two nonblocking initialization dependencies remain in the newly added tests, including the ambient-zero issue already identified by CodeRabbit. For that native case, its proposed fix also keeps torch's ambient stream different from the handle's stream during g.execute; the current nested side-stream context does not exercise that mismatch. My existing approval stands.

**[P2] Order each round's source initialization before its launch stream** ([source](https://github.com/NVIDIA/cudnn-frontend/blob/ebd29bbef4a563550cbfc045cd95cd7f3e81eefe/test/python/sdpa/test_stream_context_contract.py#L348))

The new per-round case creates source on the device's current stream, then reads it on an independent launch stream without an event/wait between them. Synchronizing after the copy only waits for both operations to finish; it cannot ensure the copy read initialized values. A controlled delay of the factory's queued fill makes the correct helper fail all 64 elements (0 instead of 3). Complete setup before entering the trial, or have launch wait for the initialization stream after the tensor factories. This is test flakiness, not a production helper defect.

Evidence: Unmodified test on L40S/A100, with the factory's fill delayed on its actual producer stream: 64/64 mismatches, max absolute error 3. The same delayed initialization with an explicit completion dependency passes. The original ordering test also passes its previous delayed-full control after this revision's setup barrier. Attribution: new per-round test repeats the previously corrected initialization dependency pattern

- **[P2] Still open:** [Order the native test's output reset before the engine write](https://github.com/NVIDIA/cudnn-frontend/pull/1149#discussion_r4056137974).

Validation: Exact-head normal run: 28 passed (13 stream/device/native-capture tests and 15 import-boundary tests), with L40S and A100 visible; cuDNN 9.25.1.1, PyTorch 2.13.0+cu130. Imported source and identical native-source provenance verified. Five controlled outcomes: original ordering test with delayed initialization passes; new per-round test with delayed initialization fails and its ordered control passes; new native case with delayed ambient zero fails (65004/65536 mismatches, max error 0.524852) and its ordered control passes. The experiment asserts all expected outcomes and exits successfully. Reviewed the complete one-file incremental diff (204 additions, 16 deletions). Production helpers and kernels remain unchanged.

Limitations: These are nonblocking test-reliability findings. No production performance comparison or broad GPU pipeline is needed for this test-only follow-up. The exact-head fork Style check is awaiting maintainer authorization; the unchanged read-only formatting workflow was inspected.

Comment only: this PR already has an active approval; preserving existing approvals.

**Comment by YangXu1990uiuc** [comment]
The [Style check for `ebd29bbe`](https://github.com/NVIDIA/cudnn-frontend/actions/runs/35489655856) passed after maintainer authorization. The follow-up review retains the existing approval and leaves the two new test-initialization dependencies as nonblocking comments; no broad GPU pipeline was requested.

Codex bot review — model `gpt-6-astra`.

**Comment by 0z5a** [comment]
Both initialization-order findings are fixed in `3483f155`:

- **per-round case**: the factory fill is recorded and the launch stream waits on that record, with the device barrier taken before the round's trial;
- **native case**: the reset is queued on the bound stream and `g.execute` is now called with `ambient` current, so `ambient != handle` is genuinely exercised -- the previous nested side-stream context had made the two equal. The capture/replay path got the same treatment (inputs updated and output reset on the capture stream, in program order ahead of the replay).

28 passed here (13 contract + 15 import-boundary). Honest caveat: the delayed-factory control passes on the fixed cases, but it also passes on the old code on this L20 -- the host schedules the cross-stream read after the producer even without the handshake, the same masking already recorded for the value-race trial. So the fix rests on your A100/L40S reproduction plus the ordering structure, not on a locally discriminating control.

**Review by YangXu1990uiuc — COMMENTED** [review]
**Codex bot review · model: `gpt-6-astra`**

Reviewed `3483f1556c89` against base `667fe4ce8ce4`.

Thank you for your contribution! Both initialization-order findings from my previous review are resolved. The per-round setup completes before the launch reads it, and the native test orders its output reset on the handle stream while calling execute with a different ambient stream. Capture/replay updates are also ordered on the capture stream. No new findings; my existing approval stands.

Validation: Exact-head tests on L40S/A100: 28 passed (13 stream/device/native-capture cases and 15 import-boundary cases), cuDNN 9.25.1.1 and PyTorch 2.13.0. Matching native-source identity verified before reusing the binding. Reused the previous review's controlled-delay probes on the same hardware: delayed per-round factory initialization and delayed native output reset now both pass, as do their explicitly ordered controls. The original ordering test also still passes with delayed initialization. These are the same two failure triggers reproduced on the previous head, not just an ordinary green suite. Reviewed the complete incremental one-file diff. Production code and workflow contents are unchanged.

Limitations: This validates helper and native-backend stream behavior; it does not claim new FROST kernel or architecture coverage. No production performance comparison or broad GPU CI run is needed for this test-only follow-up. The current-head fork Style check is being authorized separately after inspecting the unchanged read-only workflow.

Comment only: this PR already has an active approval; preserving existing approvals.

**Comment by YangXu1990uiuc** [comment]
The [Style check for `3483f155`](https://github.com/NVIDIA/cudnn-frontend/actions/runs/35491996948) passed after maintainer authorization. Both initialization-order findings are resolved in the [follow-up review](https://github.com/NVIDIA/cudnn-frontend/pull/1149#pullrequestreview-5259748495), and the existing approval stands.

Codex bot review — model `gpt-6-astra`.

**Comment by 0z5a** [comment]
Same class of fix in `21a68b0`: the dense-lane case only skipped when *every* plan was a Python engine, so on an SM120 it executed a FROST plan while claiming to test the backend one. It now selects the first non-Python plan explicitly.

SM120 results at that head (RTX 5090, cuDNN 9.24, torch 2.14.0+cu130, second venv confirmed): this file **13 passed**; the SM120 FROST suites give fwd DSL **107 passed**, split-kv **29 passed**, bwd DSL **118 passed / 1 skipped** (strided Stats needs cuDNN >= 9.26), prepared/THD **11 passed / 15 skipped** (SM100-line only).

Unrelated red on that host: `sdpa/frost/test_sdpa_fwd_fp8_sm120.py` is **6 failed / 125 passed**, all `max|O-ref|=nan` on tail geometries (`s96`, `s97`, `s110`, the 255/321 tail, and e5m2), reproducible in both venvs. This PR touches no kernel, so that looks like a pre-existing SM120 FP8 issue rather than anything in this diff.

**Review by YangXu1990uiuc — COMMENTED** [review]
**Codex bot review · model: `gpt-6-astra`**

Reviewed `21a68b0b7ed6` against base `b100e34239d4`.

Thank you for your contribution! The native-plan pin fixes the test's routing assumption without changing the stream experiment. The previous initialization ordering and changed-input capture/replay checks remain intact. No new findings; my existing approval stands.

Validation: Reviewed the complete one-file follow-up from 3483f155: 7 additions and 1 deletion. select_plan strictly preserves the native choice through build_plans; production and workflow sources are unchanged. Exact-head stream-contract module with L40S and A100 visible, Python engines enabled, DSL 4.7.0/cuDNN 9.25.1: 13 passed, including both multi-device cases. The changed native dense-stream test also passed independently on SM100, including the distinct ambient/handle streams and capture replay with changed inputs. Verified checkout imports and matching native sources before binding reuse.

Limitations: The author's SM120 result was not independently rerun on SM120. These follow-up checks validate helper/native-backend behavior and do not claim FROST kernel coverage. The unrelated SM120 FP8 failures reported in the discussion are not diagnosed by this test-only review. No production performance or broad GPU CI rerun is needed for this small test selection change. The pending current-head Style workflow is suitable for maintainer authorization.

Comment only: this PR already has an active approval; preserving existing approvals.

**Comment by YangXu1990uiuc** [comment]
The [Style check for `21a68b0b`](https://github.com/NVIDIA/cudnn-frontend/actions/runs/35503071630) passed after authorization. The follow-up review confirms the native-plan selection fix and retains the existing approval.

Codex bot review — model `gpt-6-astra`.

**Comment by 0z5a** [comment]
Independently reproduced your focused numbers at `21a68b0` on two parts here: **13 passed** on an L40S (SM89, cuDNN 9.26.0.51, torch 2.11.0+cu128, driver 560.35.03, two devices visible) and **13 passed** on an RTX 5090 (SM120, cuDNN 9.24, torch 2.14.0+cu130). The native dense-lane case selected `eng8_k24=1_k27=0_k38=0_k40=3_k41=1` on both.

One honest note on the delayed-factory control: I re-ran it on that L40S and it still does **not** discriminate there -- with the factories delayed behind a device spin, both the old code and the fixed code pass, i.e. this host (torch 2.11+cu128, driver 560) schedules the cross-stream read after the producer even without the handshake, the same masking I see on the L20. Your A100/L40S reproduction with torch 2.13+cu130 stands as the sensitivity evidence; I could not re-create it here because this instance's driver (560.35.03) caps CUDA 12.8, so the cu130 stack will not run.

**Review by YangXu1990uiuc — COMMENTED** [review]
**Codex bot review · model: `gpt-6-astra`**

Reviewed `a0e633f606bd` against base `a5d61f45c08d`.

Thank you for your contribution! The entire Git tree is identical to the previously reviewed 21a68b0b. No new findings; the existing approval stands, and the prior stream/device validation remains applicable.

Validation: Both revisions resolve to tree 55ca57410426e6ffbbe8999ca5cd6f5e0156e26f; there are no source, test or workflow changes. Reused the prior independent results on the identical code: the complete stream-contract module passed all 13 cases with L40S and A100 visible, including both multi-device cases; the native explicit-stream/capture-replay test also passed separately on SM100.

Limitations: These GPU runs were recorded at the previous SHA with the identical tree, not rerun on this history-only update. They establish the scoped helper/native-backend behavior, not broader FROST coverage or performance.

Comment only: this PR already has an active approval; preserving existing approvals.

**Comment by YangXu1990uiuc** [comment]
Authorized the pending [Style check](https://github.com/NVIDIA/cudnn-frontend/actions/runs/35962452486) for `a0e633f606bd`; it passed. No additional GPU pipeline was needed for this unchanged-tree follow-up.

Codex bot review — model `gpt-6-astra`.
